File: Algorithms/Data Structure dependent/Graphs/DFS/robotCleaner.py
# """
# This is the robot's control interface.
# You should not implement it, or speculate about its implementation
# """
#class Robot:
#    def move(self):
#    def turnRight(self):
#    def clean(self):

class Solution:
    def cleanRoom(self, robot):
        dirs = [(0,1), (1,0), (0, -1), (-1,0)]
        visited = set()
        def search(robot, visited, x, y, curDir):
            nonlocal dirs
            if (x,y) not in visited:
                robot.clean()
                visited.add((x,y))
                # print(f"{x},{y}")

                for i in range(4):
                    nextDir = (i+curDir) % 4
                    nextDirX, nextDirY = dirs[nextDir]
                    if robot.move():
                        search(robot, visited, x+nextDirX, y+nextDirY, nextDir)
                    robot.turnRight()

            # face the way we came in
            robot.turnRight()
            robot.turnRight()
            robot.move() # go back to old cell
            # face the way we were facing when we came in
            robot.turnRight()
            robot.turnRight()

        search(robot, visited, 0, 0, 0)

File: Algorithms/Data Structure dependent/Graphs/DFS/robotCleaner.cpp
typedef std::pair<int, int> pii;

struct pair_hash {
    inline std::size_t operator()(const std::pair<int,int> & v) const {
        return v.first*31+v.second;
    }
};

typedef std::unordered_set<pii, pair_hash> u_pii_set;
class Solution {
public:
    const int dirX[4] = {0, 1, 0, -1};
    const int dirY[4] = {1, 0, -1, 0};

    void goBack(Robot& robot){
        robot.turnRight();
        robot.turnRight();
        robot.move();
        robot.turnRight();
        robot.turnRight(); 
    }

    void moveRobot(Robot& robot,  u_pii_set& visited, const pii& pos, int curDir){
        auto nextPos = make_pair(pos.first+ dirX[curDir], pos.second + dirY[curDir]);
        cleanRoom(robot, visited, nextPos, curDir);

    }

    void cleanRoom(Robot& robot,  u_pii_set& visited, const pii& pos, int curDir = 0){
        if(visited.find(pos) == visited.end()){
            robot.clean();
            visited.insert(pos);
            
            /* Go straight */
            if(robot.move()) moveRobot(robot, visited, pos, curDir);

            /* Go right */
            robot.turnRight();
            if(robot.move()) moveRobot(robot, visited, pos, (curDir + 1) % 4);
            
            /* Go back*/
            robot.turnRight();
            if(robot.move()) moveRobot(robot, visited, pos, (curDir + 2) % 4);

            /* Go left */
            robot.turnRight();
            if(robot.move()) moveRobot(robot, visited, pos, (curDir + 3) % 4);

            /* Re straighten */
            robot.turnRight();
        }

        /* Go back and face same direction */
        goBack(robot);
    }

    void cleanRoom(Robot& robot){
        u_pii_set visited;
        cleanRoom(robot, visited, make_pair(0, 0));
    }
};